### PR TITLE
Bring pullbacks of divisors to the best state.

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
With this version I was able to pull back the complete algebraic lattice of a resolved K3 surface along some endomorphism and identify the components on the other end; on a local machine with 8GB RAM. 

The corresponding [test file](https://github.com/HechtiDerLachs/Oscar.jl/blob/elliptic_surface_mod/vinberg-2.jl) can be found on this [branch](https://github.com/HechtiDerLachs/Oscar.jl/tree/elliptic_surface_mod). 

We make some random choices in the computations and it's not yet clear how reliably the pullback works in practice. But I got through on the first try.

### CODE REVIEW

File: experimental/Schemes/RationalMap.jl [hunk 1]
```diff
@@ -535,70 +544,108 @@ function pullback(phi::RationalMap, C::AbsAlgebraicCycle)
...
+  @show all_V
+  V = first(all_V)
+  @show findfirst(x->x===V, affine_charts(Y))
+
+  all_U = copy(affine_charts(X))
+  function complexity(U::AbsSpec)
+    a = realization_preview(phi, U, V)
+    return maximum(vcat([total_degree(numerator(f)) for f in a], [total_degree(denominator(f)) for f in a]))
+  end
+  sort!(all_U, lt=(x,y)->complexity(x)<complexity(y))
```
Comment (fingolfin) [inline experimental/Schemes/RationalMap.jl]: This should also work (untested)
```suggestion
  sort!(all_U, by=complexity)
```

File: experimental/Schemes/RationalMap.jl [hunk 2]
```diff
@@ -535,70 +544,108 @@ function pullback(phi::RationalMap, C::AbsAlgebraicCycle)
...
-        JJ = IdealSheaf(X, domain(psi), gens(J))
-        comps[JJ] = C[I]
-        success = true
-        break
-      end
+    J = pullback(psi)(saturated_ideal(I(V)))
+    if !isone(J)
+      JJ = IdealSheaf(X, domain(psi), gens(J))
+      return JJ
+      break
```
Comment (fingolfin) [inline experimental/Schemes/RationalMap.jl]:
```suggestion
```

File: experimental/Schemes/RationalMap.jl [hunk 3]
```diff
@@ -535,70 +544,108 @@ function pullback(phi::RationalMap, C::AbsAlgebraicCycle)
...
-        cod_ideal = cod_ideal + ideal(OO(U), complement_equation(domain(psi)))
+function _pullback(phi::RationalMap, I::IdealSheaf)
+  X = domain(phi)
+  Y = codomain(phi)
+  scheme(I) === Y || error("ideal sheaf not defined on the correct scheme")
+  # Find a patch in Y on which this component is visible
+  all_V = [V for V in affine_charts(Y) if !isone(I(V))]
+
+  @show all_V
+  min_var = minimum([ngens(OO(V)) for V in all_V])
```
Comment (fingolfin) [inline experimental/Schemes/RationalMap.jl]: Clearly you have bigger fish to fry, but let me just point out that code like the following is always inefficient, as you create a vector to store all the `ngens` values, when you really only need to store a single value at any moment.
```suggestion
  min_var = minimum(ngens(OO(V)) for V in all_V)
```

File: experimental/Schemes/RationalMap.jl [hunk 4]
```diff
@@ -535,70 +544,108 @@ function pullback(phi::RationalMap, C::AbsAlgebraicCycle)
...
+  for U in all_U
+    psi_loc = realize_maximally_on_open_subset(phi, U, V)
+    # If we are in different components, skip
+    length(psi_loc) > 0 || continue
+    found = 0
+    J = ideal(OO(domain(first(psi_loc))), elem_type(OO(domain(first(psi_loc))))[])
+    cod_ideal = ideal(OO(U), elem_type(OO(U))[])
+    for (k, psi) in enumerate(psi_loc)
+      if dim(cod_ideal) < dim(I)
+        found = 0
```
Comment (fingolfin) [inline experimental/Schemes/RationalMap.jl]:
```suggestion
```

File: experimental/Schemes/RationalMap.jl [hunk 5]
```diff
@@ -535,70 +544,108 @@ function pullback(phi::RationalMap, C::AbsAlgebraicCycle)
...
+    a = realization_preview(phi, U, V)
+    return maximum(vcat([total_degree(numerator(f)) for f in a], [total_degree(denominator(f)) for f in a]))
+  end
+  sort!(all_U, lt=(x,y)->complexity(x)<complexity(y))
+
+  for U in all_U
+    psi_loc = realize_maximally_on_open_subset(phi, U, V)
+    # If we are in different components, skip
+    length(psi_loc) > 0 || continue
+    found = 0
```
Comment (fingolfin) [inline experimental/Schemes/RationalMap.jl]:
```suggestion
```

File: experimental/Schemes/RationalMap.jl [hunk 6]
```diff
@@ -510,20 +510,29 @@ function pullback(phi::RationalMap, C::AbsAlgebraicCycle)
...
-  # deg_bound = minimum([maximum([total_degree(lifted_numerator(g)) for g in gens(I(V))]) for V in all_V])
-  # @show deg_bound
-  # all_V = [V for V in all_V if minimum([total_degree(lifted_numerator(g)) for g in gens(I(V))]) == deg_bound]
-  # @show all_V
-  # V = first(all_V)
-  # @show findfirst(x->x===V, affine_charts(Y))
+    @show "trying cheap"
+    pbI = _try_pullback_cheap(phi, I)
+    if pbI === nothing
+      @show "trying randomized"
```
Comment (simonbrandhorst) [inline experimental/Schemes/RationalMap.jl]: These verbose printing messages are great. Here is the way to do it properly:
```suggestion
      @vprint :RationalMap 1 "trying randomized pullback\n"
```

Comment (simonbrandhorst) [reply experimental/Schemes/RationalMap.jl]: Then one also has to add an `add_verbose_scope!(:RationalMap)` in our `__init__` in `Oscar.jl`

File: experimental/Schemes/RationalMap.jl [hunk 7]
```diff
@@ -510,20 +510,29 @@ function pullback(phi::RationalMap, C::AbsAlgebraicCycle)
...
-  # @show min_var
-  # all_V = [V for V in all_V if ngens(OO(V)) == min_var]
-  # @show all_V
-  # deg_bound = minimum([maximum([total_degree(lifted_numerator(g)) for g in gens(I(V))]) for V in all_V])
-  # @show deg_bound
-  # all_V = [V for V in all_V if minimum([total_degree(lifted_numerator(g)) for g in gens(I(V))]) == deg_bound]
-  # @show all_V
-  # V = first(all_V)
-  # @show findfirst(x->x===V, affine_charts(Y))
+    @vprint :RationalMap 1 "trying cheap pullback\n"
```
Comment (lgoettgens) [inline experimental/Schemes/RationalMap.jl:513]:
```suggestion
    @vprintln :RationalMap 1 "trying cheap pullback"
```
This is even easier to read

File: experimental/Schemes/RationalMap.jl [hunk 8]
```diff
@@ -535,70 +548,102 @@ function pullback(phi::RationalMap, C::AbsAlgebraicCycle)
...
+    pullbacks = IdDict{AbsSpec, Ideal}()
     for U in all_U
       psi = cheap_realization(phi, U, V)
       U_sub = domain(psi)
-      J = pullback(psi)(saturated_ideal(I(V)))
+      pullbacks[U] = pullback(psi)(saturated_ideal(I(V)))
+    end
+      #J = pullback(psi)(saturated_ideal(I(V)))
+    function new_complexity(U::AbsSpec)
+      return sum(total_degree.(lifted_numerator(gens(pullbacks[U]))); init=0)
```
Comment (simonbrandhorst) [inline experimental/Schemes/RationalMap.jl:559]:
```suggestion
      return sum(total_degree.(lifted_numerator(gens.(pullbacks[U]))); init=0)
```
from the looks of it. But I did not confirm with an example

**Comment by simonbrandhorst** [comment]
Looks o.k. for me. But some more comments in the code as to what is going on would be nice. 
This scheme code is really hard to read. Is supposed to be a draft still?

**Comment by simonbrandhorst** [comment]
Could the random realizations cause random failures in our CI? 
If yes, please avoid it somehow (maybe by working over GF(2)?)

File: experimental/Schemes/RationalMap.jl (hunk 8 above)
Comment (simonbrandhorst) [reply experimental/Schemes/RationalMap.jl:559]: o.k. nevermind nonsense

**Review by simonbrandhorst — APPROVED** [review]

**Pull request merged**